File: DSA/String/Code/CheckForSame.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <algorithm>
using namespace std;

// Check for Same (String Comparison) in C++
// Different methods to compare two strings

int main() {
    string str1 = "Hello";
    string str2 = "Hello";
    string str3 = "hello";
    string str4 = "World";
    
    // Method 1: Using == operator for string objects
    cout << "Using == operator:\n";
    cout << "(str1 == str2): " << (str1 == str2) << "\n";  // 1 = true
    cout << "(str1 == str3): " << (str1 == str3) << "\n";  // 0 = false
    cout << "(str1 == str4): " << (str1 == str4) << "\n";  // 0 = false
    
    // Method 2: Using compare() method
    cout << "\nUsing compare() method:\n";
    cout << "str1.compare(str2): " << str1.compare(str2) << "\n";
    cout << "str1.compare(str3): " << str1.compare(str3) << "\n";
    
    // Method 3: Case-insensitive comparison
    cout << "\nCase-insensitive comparison:\n";
    string lower1 = str1;
    string lower3 = str3;
    transform(lower1.begin(), lower1.end(), lower1.begin(), ::tolower);
    transform(lower3.begin(), lower3.end(), lower3.begin(), ::tolower);
    cout << "Lower case equal: " << (lower1 == lower3) << "\n";
    
    // Method 4: Comparing first n characters
    cout << "\nComparing first 3 characters:\n";
    string str5 = "Helloworld";
    string str6 = "Hellouni";
    cout << "substr comparison: " << (str5.substr(0, 3) == str6.substr(0, 3)) << "\n";
    
    // Method 5: C-style string comparison with strcmp()
    cout << "\nC-style string comparison:\n";
    const char *cstr1 = "Hello";
    const char *cstr2 = "Hello";
    cout << "strcmp result: " << strcmp(cstr1, cstr2) << "\n";
    
    return 0;
}

File: DSA/String/Code/CheckForSame.py
# Check for Same (String Comparison) in Python
# Different methods to compare two strings

str1 = "Hello"
str2 = "Hello"
str3 = "hello"
str4 = "World"

# Method 1: Using == operator
print("Using == operator:")
print(f"str1 == str2: {str1 == str2}")  # True
print(f"str1 == str3: {str1 == str3}")  # False
print(f"str1 == str4: {str1 == str4}")  # False

# Method 2: Using != operator
print("\nUsing != operator:")
print(f"str1 != str3: {str1 != str3}")  # True
print(f"str1 != str2: {str1 != str2}")  # False

# Method 3: Case-insensitive comparison
print("\nCase-insensitive comparison:")
print(f"str1.lower() == str3.lower(): {str1.lower() == str3.lower()}")  # True
print(f"str1.upper() == str3.upper(): {str1.upper() == str3.upper()}")  # True

# Method 4: Comparing specific parts
print("\nComparing first 3 characters:")
str5 = "Helloworld"
str6 = "Hellouni"
print(f"str5[:3] == str6[:3]: {str5[:3] == str6[:3]}")  # True

# Method 5: Ignoring case and spaces
print("\nIgnoring case and spaces:")
str7 = "Hello World"
str8 = "hello world"
print(f"Normalized equal: {str7.lower().strip() == str8.lower().strip()}")

# Method 6: Using == vs is
print("\nUsing == vs is (identity vs equality):")
str9 = "Hello"
str10 = "Hello"
str11 = str9
print(f"str9 == str10: {str9 == str10}")  # True (value equal)
print(f"str9 is str10: {str9 is str10}")  # May be True (same object)
print(f"str9 is str11: {str9 is str11}")  # True (same object)

# Method 7: Using comparison operators
print("\nUsing comparison operators:")
print(f"'apple' < 'banana': {'apple' < 'banana'}")  # True
print(f"'zebra' > 'apple': {'zebra' > 'apple'}")    # True
